# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1429294020 14400
#      Fri Apr 17 14:07:00 2015 -0400
# Node ID c58cf3a6551934a8f4fb90b4fc977f915e6f602d
# Parent  4d0c7fec0a315c2c551488acff2d7a3ff61a42db
improve copy to clipboard code for Qt plotting

* Figure.cc (Figure::copy_figure_callback): Append format to temporary
file name.  Check that QImage is valid before copying to clipboard.

diff --git a/libgui/graphics/Figure.cc b/libgui/graphics/Figure.cc
--- a/libgui/graphics/Figure.cc
+++ b/libgui/graphics/Figure.cc
@@ -46,16 +46,17 @@ along with Octave; see the file COPYING.
 #include "Container.h"
 #include "Figure.h"
 #include "FigureWindow.h"
 #include "MouseModeActionGroup.h"
 #include "QtHandlesUtils.h"
 
 #include "file-ops.h"
 #include "unwind-prot.h"
+#include "utils.h"
 
 #include "octave-qt-link.h"
 
 #include "builtin-defun-decls.h"
 
 namespace QtHandles
 {
 
@@ -823,33 +824,41 @@ delete_file (const std::string& file)
   
 void
 Figure::copy_figure_callback (const std::string& format)
 {
   std::string msg;
 
   unwind_protect frame;
 
-  std::string file = octave_tempnam ("", "oct-", msg);
+  std::string file = octave_tempnam ("", "oct-", msg) + "." + format;
 
   if (file.empty ())
     {
-      // FIXME: report error contained in message.
+      // Report error?
       return;
     }
 
   frame.add_fcn (delete_file, file);
 
   std::string device = "-d" + format;
 
   Ffeval (ovl ("print", file, device));
 
   QClipboard *clipboard = QApplication::clipboard ();
 
-  clipboard->setImage (QImage (file.c_str ()));
+  QImage img (file.c_str (), format.c_str ());
+
+  if (img.isNull ())
+    {
+      // Report error?
+      return;
+    }
+
+  clipboard->setImage (img);
 }
   
 void
 Figure::fileSaveFigureAs (void)
 {
   fileSaveFigure (true);
 }
 
